File: ros/src/npu_subscriber.cpp
/*
 * C++ 订阅者：在 macOS 端订阅 QEMU guest NPU 节点发布的 /npu/result。
 * 收到消息后在回调里加业务逻辑即可。
 */
#include <rclcpp/rclcpp.hpp>
#include "npu_ros/msg/npu_result.hpp"
#include <functional>

using npu_ros::msg::NpuResult;

class NpuSubscriber : public rclcpp::Node
{
public:
    NpuSubscriber() : Node("npu_subscriber")
    {
        sub_ = this->create_subscription<NpuResult>(
            "/npu/result", 10,
            std::bind(&NpuSubscriber::on_result, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "订阅者已启动，监听 /npu/result");
    }

private:
    /* 收到 NPU 推理结果时的回调：在这里加你的业务逻辑 */
    void on_result(const NpuResult::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(),
            "收到推理结果: class=%d label=%s scores=[%d,%d,%d] input=[%d,%d,%d,%d]",
            msg->class_id, msg->label.c_str(),
            msg->scores[0], msg->scores[1], msg->scores[2],
            msg->input[0], msg->input[1], msg->input[2], msg->input[3]);

        // TODO: 在这里根据 msg 做后续处理，例如：
        //   if (msg->class_id == 2) { RCLCPP_WARN(...); /* 触发告警 */ }
        //   转发到别的 topic / 写日志 / 调外部服务 ...
    }

    rclcpp::Subscription<NpuResult>::SharedPtr sub_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<NpuSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: ros/scripts/npu_subscriber.py
#!/usr/bin/env python3
"""Python 订阅者：在 macOS 端订阅 QEMU guest NPU 节点发布的 /npu/result。"""
import rclpy
from rclpy.node import Node
from npu_ros.msg import NpuResult


class NpuSubscriber(Node):
    def __init__(self):
        super().__init__('npu_subscriber')
        self.subscription = self.create_subscription(
            NpuResult, '/npu/result', self.on_result, 10)
        self.get_logger().info('订阅者已启动，监听 /npu/result')

    def on_result(self, msg: NpuResult):
        """收到推理结果时的回调：在这里加你的业务逻辑。"""
        self.get_logger().info(
            f'收到推理结果: class={msg.class_id} label={msg.label} '
            f'scores={list(msg.scores)} input={list(msg.input)}')

        # TODO: 在这里根据 msg 做后续处理，例如：
        #   if msg.class_id == 2:
        #       self.get_logger().warn('检测到 critical，触发告警！')


def main():
    rclpy.init()
    node = NpuSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
